File: 778.swim-in-rising-water.cpp
/*
 * @lc app=leetcode id=778 lang=cpp
 *
 * [778] Swim in Rising Water
 */
#include <bits/stdc++.h>

using std::array;
using std::greater;
using std::priority_queue;
using std::vector;

// @lc code=start
class Solution {
    static constexpr int D[] {-1, 0, 1, 0, -1};

public:
    // T: O(n ^ 2 * log(n)) because log(n ^ 2) = 2log(n)
    // S: O(n ^ 2)
    // (For a more generalized solution, see the java version)
    int swimInWater(const vector<vector<int>>& grid) {
        const int n = grid.size();

        // priority queue stores tuples (grid[r][c], r, c)
        priority_queue<array<int, 3>, vector<array<int, 3>>, greater<array<int, 3>>> min_heap;
        vector<vector<uint8_t>> vis(n, vector<uint8_t>(n, 0));

        min_heap.push({grid[0][0], 0, 0});
        vis[0][0] = 1;
        
        int ans = grid[n - 1][n - 1];  // the final ans must be no less than the time value of the destination cell
        while (min_heap.size()) {
            auto [t, r, c] = min_heap.top(); min_heap.pop();  // visit the cell with minimum t first
            ans = std::max(ans, t);

            for (int d = 0; d < 4; ++d) {
                int nr = r + D[d], nc = c + D[d + 1];
                if (nr == n - 1 && nc == n - 1) return ans;  // IMPORTANT: we can return here because ans is initialized to grid[-1][-1]
                if (0 <= nr && nr < n && 0 <= nc && nc < n && !vis[nr][nc]) {
                    min_heap.push({grid[nr][nc], nr, nc});
                    vis[nr][nc] = 1;
                }
            }
        }
        return ans;
    }
};
// @lc code=end

File: 778.swim-in-rising-water.java
/*
 * @lc app=leetcode id=778 lang=java
 *
 * [778] Swim in Rising Water
 */
import java.util.PriorityQueue;

// @lc code=start
class Solution {
    private static final int[] D = {1, 0, -1, 0, 1};

    // Simplified Dijkstra's algorithm
    // T: O(m * n * log(m * n))
    // S: O(m * n)
    // (this is more generalized than the cpp version)
    public int swimInWater(int[][] grid) {
        final int m = grid.length, n = grid[0].length;

        // With custom comparator, it's slightly more memory-efficient than storing tuples (grid[r][c], r, c)
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((p1, p2) -> Integer.compare(grid[p1[0]][p1[1]], grid[p2[0]][p2[1]]));
        boolean[][] vis = new boolean[m][n];
        
        minHeap.offer(new int[] {0, 0});
        vis[0][0] = true;

        int ans = 0;
        while (!minHeap.isEmpty()) {
            var cur = minHeap.poll();
            ans = Math.max(ans, grid[cur[0]][cur[1]]);
            if (cur[0] == m - 1 && cur[1] == n - 1) break;

            for (int d = 0; d < 4; ++d) {
                int r = cur[0] + D[d], c = cur[1] + D[d + 1];
                if (0 <= r && r < m && 0 <= c && c < n && !vis[r][c]) {
                    minHeap.offer(new int[] {r, c});
                    vis[r][c] = true;
                }
            }
        }

        return ans;
    }
}
// @lc code=end

